File: geekbang/algorithm/source-code/3/lc236 二叉树的最近公共祖先.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        self.father = {}
        self.calcFather(root)

        redNodes = {}
        redNodes[root.val] = True
        while p.val != root.val:
            redNodes[p.val] = True
            p = self.father[p.val]
        while q.val not in redNodes:
            q = self.father[q.val]
        return q   
    
    def calcFather(self, root):
        if not root:
            return
        if root.left:
            self.father[root.left.val] = root
            self.calcFather(root.left)
        if root.right:
            self.father[root.right.val] = root
            self.calcFather(root.right)

File: geekbang/algorithm/source-code/3/lc236 二叉树的最近公共祖先.go
func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	// 递归终止条件1：如果当前节点是空，直接返回nil
	if root == nil {
		return nil
	}
	// 递归终止条件2：如果当前节点等于p或者q，直接找到返回root
	if root.Val == p.Val || root.Val == q.Val {
		return root
	}
	left := lowestCommonAncestor(root.Left, p, q)
	right := lowestCommonAncestor(root.Right, p, q)
	// 左右子节点找到了，直接返回root
	if left != nil && right != nil {
		return root
	}
	// 左边到头了，返回right
	if left == nil {
		return right
	}
	// 否则返回left
	return left
}